### PR TITLE
=str #25045 adding Java/Scala interop to SourceQueue and SinkQueue

### PR DESCRIPTION
**Pull request by svezfaz**
Handles #25045

### CODE REVIEW

**Comment by johanandren** [comment]
I think some of those are actual problems that we cannot just filter, the removal of `SourceQueue` I think cannot be safely done as it is a public API

**Comment by svezfaz** [comment]
@johanandren correct, I went a bit overboard with the cleanup there.
It's annoying as it doesn't really look like `SourceQueue` and `SinkQueue` are really needed as separate traits.. thoughts to be kept for a major release I guess.

**Review by johanandren — CHANGES_REQUESTED** [review]
Looks mostly good, just one small thing with the test case

**Comment by johanandren** [comment]
Heh, ups reviewed in the wrong tab here, ignore my last comment :)

**Comment by johanandren** [comment]
And I agree, it is annoying but not much we can do.

**Comment by svezfaz** [comment]
@johanandren these two were internal API classes, and they should be safe to remove?
```
[error]  * class akka.stream.impl.SinkQueueAdapter does not have a correspondent in current version
[error]    filter with: ProblemFilters.exclude[MissingClassProblem]("akka.stream.impl.SinkQueueAdapter")
[error]  * class akka.stream.impl.SourceQueueAdapter does not have a correspondent in current version
[error]    filter with: ProblemFilters.exclude[MissingClassProblem]("akka.stream.impl.SourceQueueAdapter")
```

**Comment by raboof** [comment]
@svezfaz yes, since it's a final class and internal API those should be OK to ignore.

**Comment by johanandren** [comment]
Still some bin comp issues, note that you can run the check locally like so `akka-stream/mimaReportBinaryIssues` to avoid the turnaround cycle with push-wait-for-ci-server.

**Comment by svezfaz** [comment]
I have added MiMa filters for `ReversedMissingMethodProblem`. These are for the new methods.

```
+ProblemFilters.exclude[ReversedMissingMethodProblem]("akka.stream.javadsl.SourceQueue.asScala")
+ProblemFilters.exclude[ReversedMissingMethodProblem]("akka.stream.javadsl.SourceQueueWithComplete.asScala")
+ProblemFilters.exclude[ReversedMissingMethodProblem]("akka.stream.javadsl.SinkQueue.asScala")
+ProblemFilters.exclude[ReversedMissingMethodProblem]("akka.stream.javadsl.SinkQueueWithCancel.asScala")
+ProblemFilters.exclude[ReversedMissingMethodProblem]("akka.stream.scaladsl.SourceQueue.asJava")
+ProblemFilters.exclude[ReversedMissingMethodProblem]("akka.stream.scaladsl.SourceQueueWithComplete.asJava")
+ProblemFilters.exclude[ReversedMissingMethodProblem]("akka.stream.scaladsl.SinkQueue.asJava")
+ProblemFilters.exclude[ReversedMissingMethodProblem]("akka.stream.scaladsl.SinkQueueWithCancel.asJava")
```

**Comment by johanandren** [comment]
Hmm, this means that if anyone implemented either of those traits, that will fail, so I don't think we can add those filters. This would only be safe for certain if we had marked the trait as not to be implemented by users, however unlikely a user could have implemented their own subtype with that method that will now break things.

The only safe way I can think of is making the `toScala`/`toJava` methods on the companion objects of the respective traits instead. But then I think there is an additional complication for the Java side with companion objects of traits not being accessible from Java or something like that, I never quite remember correctly.

**Comment by svezfaz** [comment]
@johanandren I have reworked it to put the relevant functions in the companion objects

**Review by johanandren — APPROVED** [review]
LGTM, thanks!

File: akka-stream/src/main/scala/akka/stream/scaladsl/Queue.scala [hunk 1]
```diff
@@ -40,36 +47,60 @@ trait SourceQueue[T] {
...
+   * - failed when
+   *   - the stage faces an internal failure
+   *   - the [[SourceQueueWithComplete.fail]] method is invoked.
    */
   def watchCompletion(): Future[Done]
 }
 
+object SourceQueueWithComplete {
+  implicit def QueueOps[T](f: SourceQueueWithComplete[T]): QueueOps[T] = new QueueOps[T](f)
+  final class QueueOps[T](val queue: SourceQueueWithComplete[T]) extends AnyVal {
```
Comment (patriknw) [inline akka-stream/src/main/scala/akka/stream/scaladsl/Queue.scala]: use `implicit class` instead?

File: akka-stream/src/main/scala/akka/stream/scaladsl/Queue.scala [hunk 2]
```diff
@@ -40,36 +47,60 @@ trait SourceQueue[T] {
...
+    def asJava: akka.stream.javadsl.SourceQueueWithComplete[T] = SourceQueueConverter.asJava(queue)
```
Comment (patriknw) [inline akka-stream/src/main/scala/akka/stream/scaladsl/Queue.scala]: why is this added as an implicit instead of just adding it to `SourceQueueWithComplete`?

File: akka-stream/src/main/scala/akka/stream/scaladsl/Queue.scala [hunk 3]
```diff
@@ -40,36 +47,60 @@ trait SourceQueue[T] {
...
+  }
+}
+
+object SourceQueueConverter {
```
Comment (patriknw) [inline akka-stream/src/main/scala/akka/stream/scaladsl/Queue.scala]: should this be internal?

**Review by patriknw — COMMENTED** [review]
some questions...

File: akka-stream/src/main/scala/akka/stream/javadsl/Queue.scala [hunk 4]
```diff
@@ -42,36 +47,60 @@ trait SourceQueue[T] {
...
+   * Method returns a [[CompletionStage]] that will be
+   * - completed if the stream completes
+   * - failed when
+   *   - the stage faces an internal failure
+   *   - the [[SourceQueueWithComplete.fail]] method is invoked.
+   */
+  def watchCompletion(): CompletionStage[Done]
+}
+
+object SourceQueueWithComplete {
```
[the remark this answers is not in the document]
Comment (svezfaz) [reply akka-stream/src/main/scala/akka/stream/javadsl/Queue.scala]: You're correct, the javadsl doesn't work.
The problem I'm having here is that I cannot add a method to `SourceQueueWithComplete` (and the other traits) not to break binary compat. Any suggestion around this?

File: akka-stream/src/main/scala/akka/stream/scaladsl/Queue.scala (hunk 2 above)
Comment (svezfaz) [reply akka-stream/src/main/scala/akka/stream/scaladsl/Queue.scala]: See comment above, cannot add methods to `SourceQueueWithComplete` not to break API. Any chance we can add this change to a future, major release? I don't think there is any hurry around this change anyway..

**Comment by patriknw** [comment]
Looks like this only requires some minor touchup, I'll try to fix that.

**Comment by patriknw** [comment]
Completed in https://github.com/akka/akka/pull/26729
